# A/B Testing Analysis Project

This project analyzes an A/B test experiment to compare the performance of a control page and a treatment page.

The main objective is to understand whether the new page improved user conversion compared to the old page.

This project uses Python, SQL, MySQL, statistical testing, and Power BI.

In [27]:
# !/usr/bin/env python3
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine


In [28]:
# create SQLAlchemy engine to connect to MySQL
engine = create_engine("mysql+pymysql://root:root1234@localhost:3306/fact_ab_test")

In [29]:
df_ab = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/A:B Test Project/Data_Raw/ab_test.csv')
df_country = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/A:B Test Project/Data_Raw/countries_ab.csv')

In [30]:
df_ab.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   id         294478 non-null  int64 
 1   time       294478 non-null  object
 2   con_treat  294478 non-null  object
 3   page       294478 non-null  object
 4   converted  294478 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 11.2+ MB


In [31]:
df_ab['time'].head(10)

0    11:48.6
1    01:45.2
2    55:06.2
3    28:03.1
4    52:26.2
5    20:49.1
6    26:46.9
7    48:29.5
8    58:09.0
9    11:06.6
Name: time, dtype: object

In [32]:
# time  change to datetime
df_ab['time_duration'] = pd.to_timedelta('00:' + df_ab['time'])

### Why add `00:` before the time column?

Our dataset time values look like this:

```python
55:06.2

In [33]:
# convert time duration to minutes
df_ab['minutes'] = df_ab['time_duration'].dt.total_seconds() / 60

In [34]:
# null value check
df_ab.isnull().sum()

id               0
time             0
con_treat        0
page             0
converted        0
time_duration    0
minutes          0
dtype: int64

In [35]:
# duplication check
df_ab.duplicated().sum()

np.int64(0)

In [36]:
# Load the cleaned A/B testing DataFrame into MySQL
df_ab_clean = df_ab.copy()

df_ab_clean.to_sql("ab_test",engine,if_exists="replace",index=False)

/var/folders/2g/g8533pvs3jnd_p49s1yblz9w0000gn/T/ipykernel_2080/3524465702.py:4: UserWarning: the 'timedelta' type is not supported, and will be written as integer values (ns frequency) to the database.
  df_ab_clean.to_sql("ab_test",engine,if_exists="replace",index=False)


294478

In [37]:
pd.read_sql('SELECT * FROM ab_test LIMIT 10', con=engine)

,id,time,con_treat,page,converted,time_duration,minutes
0,851104,11:48.6,control,old_page,0,708600000000,11.810000
1,804228,01:45.2,control,old_page,0,105200000000,1.753333
2,661590,55:06.2,treatment,new_page,0,3306200000000,55.103333
3,853541,28:03.1,treatment,new_page,0,1683100000000,28.051667
4,864975,52:26.2,control,old_page,1,3146200000000,52.436667
5,936923,20:49.1,control,old_page,0,1249100000000,20.818333
6,679687,26:46.9,treatment,new_page,1,1606900000000,26.781667
7,719014,48:29.5,control,old_page,0,2909500000000,48.491667
8,817355,58:09.0,treatment,new_page,1,3489000000000,58.150000
9,839785,11:06.6,treatment,new_page,1,666600000000,11.110000


In [38]:
# Check the structure of the country DataFrame
df_country.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290584 entries, 0 to 290583
Data columns (total 2 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   id       290584 non-null  int64 
 1   country  290584 non-null  object
dtypes: int64(1), object(1)
memory usage: 4.4+ MB


In [39]:
# null value check
df_country.isnull().sum()

id         0
country    0
dtype: int64

In [40]:
# duplication check
df_country.duplicated().sum()

np.int64(0)

In [41]:
# Load cleaned country DataFrame into MySQL
df_country_clean = df_country.copy()

df_country_clean.to_sql(
    "countries_ab",
    engine,
    if_exists="replace",
    index=False
)

290584

In [42]:
# veify the data in MySQL
pd.read_sql('SELECT * FROM countries_ab LIMIT 10', con=engine)

,id,country
0,834778,UK
1,928468,US
2,822059,UK
3,711597,UK
4,710616,UK
5,909908,UK
6,811617,US
7,938122,US
8,887018,US
9,820683,US


In [43]:
# Perform SQL query to calculate conversion rates by treatment group
pd.read_sql("""
SELECT
    con_treat,
    COUNT(*) AS total_users,
    SUM(converted) AS converted_users,
    ROUND(AVG(converted) * 100, 2) AS conversion_rate
FROM ab_test
GROUP BY con_treat;
""", engine)

,con_treat,total_users,converted_users,conversion_rate
0,control,147202,17723.0,12.04
1,treatment,147276,17514.0,11.89


## Conversion Rate Analysis by Treatment Group

This query calculates the overall conversion performance for the control and treatment groups in the A/B testing experiment.

COUNT(*) calculates the total number of users in each group.

SUM(converted) calculates the total number of converted users because:
- 1 = converted
- 0 = not converted

AVG(converted) automatically calculates the conversion rate.

The conversion rate is multiplied by 100 and rounded to 2 decimal places to display a percentage.

GROUP BY con_treat separates the results into:
- control group
- treatment group

This analysis helps evaluate whether the treatment version improved user conversion compared to the control version.

In [44]:
# Perform SQL query to calculate conversion rates by treatment group and country
pd.read_sql("""
SELECT
    c.country,
    a.con_treat,
    COUNT(*) AS total_users,
    SUM(a.converted) AS converted_users,
    ROUND(AVG(a.converted) * 100, 2) AS conversion_rate
FROM ab_test a
JOIN countries_ab c
ON a.id = c.id
GROUP BY c.country, a.con_treat
ORDER BY conversion_rate DESC;
""", engine)

,country,con_treat,total_users,converted_users,conversion_rate
0,UK,treatment,36578,4425.0,12.10
1,US,control,103059,12424.0,12.06
2,UK,control,36841,4428.0,12.02
3,CA,control,7302,871.0,11.93
4,US,treatment,103305,12257.0,11.86
5,CA,treatment,7393,832.0,11.25


## Conversion Rate Analysis by Treatment Group and Country

This query analyzes conversion performance across different countries and treatment groups.

The ab_test table contains:
- treatment group information
- conversion results

The countries_ab table contains:
- country information

An INNER JOIN is used to combine both tables using the id column.

COUNT(*) calculates the total number of users in each country and treatment group.

SUM(a.converted) calculates the total number of converted users.

AVG(a.converted) calculates the conversion rate because:
- 1 = converted
- 0 = not converted

The result is multiplied by 100 and rounded to 2 decimal places to display a percentage.

GROUP BY c.country, a.con_treat separates the analysis by:
- country
- treatment group

ORDER BY conversion_rate DESC sorts the results from highest to lowest conversion rate.

This analysis helps identify whether conversion performance differs across countries and whether the treatment page performs better in specific regions.

In [45]:
# Perform SQL query to analyze user engagement (average minutes spent) by conversion status
pd.read_sql("""
SELECT
    con_treat,
    ROUND(AVG(minutes), 2) AS avg_minutes,
    ROUND(MAX(minutes), 2) AS max_minutes,
    ROUND(MIN(minutes), 2) AS min_minutes
FROM ab_test
GROUP BY con_treat;
""", engine)


,con_treat,avg_minutes,max_minutes,min_minutes
0,control,30.08,60.0,0.0
1,treatment,30.03,60.0,0.0


## User Engagement Analysis by Treatment Group

This query analyzes user engagement by measuring the amount of time users spent on the platform in each treatment group.

AVG(minutes) calculates the average session duration for users in each group.

MAX(minutes) identifies the highest session duration recorded.

MIN(minutes) identifies the lowest session duration recorded.

ROUND(..., 2) is used to display the results with 2 decimal places for better readability.

GROUP BY con_treat separates the analysis into:
- control group
- treatment group

This analysis helps determine whether users in one group spent more time on the platform than the other group.

User engagement metrics such as session duration are important in product analytics because they can help evaluate user experience and platform interaction quality.

In [46]:
# Perform SQL query to analyze user engagement (average minutes spent) by conversion status
pd.read_sql("""
SELECT
    converted,
    ROUND(AVG(minutes), 2) AS avg_minutes,
    COUNT(*) AS total_users
FROM ab_test
GROUP BY converted;
""", engine)


,converted,avg_minutes,total_users
0,0,30.06,259241
1,1,30.00,35237


## User Engagement Analysis by Conversion Status

This query analyzes whether user engagement time differs between converted and non-converted users.

The converted column contains:
- 1 = converted user
- 0 = non-converted user

AVG(minutes) calculates the average amount of time users spent on the platform.

COUNT(*) calculates the total number of users in each conversion category.

GROUP BY converted separates the analysis into:
- converted users
- non-converted users

This analysis helps evaluate whether spending more time on the platform increases the likelihood of conversion.

Understanding engagement behaviour is important in product analytics because it helps identify whether user activity and session duration influence business outcomes such as sign-ups, purchases, or subscriptions.

Users who converted and users who did not convert spent nearly the same amount of time on the platform.

In [47]:
# Perform SQL query to analyze conversion rates by page and treatment group
pd.read_sql("""
SELECT
    page,
    con_treat,
    COUNT(*) AS total_users,
    SUM(converted) AS converted_users,
    ROUND(AVG(converted) * 100, 2) AS conversion_rate
FROM ab_test
GROUP BY page, con_treat
ORDER BY conversion_rate DESC;
""", engine)


,page,con_treat,total_users,converted_users,conversion_rate
0,old_page,treatment,1965,250.0,12.72
1,new_page,control,1928,234.0,12.14
2,old_page,control,145274,17489.0,12.04
3,new_page,treatment,145311,17264.0,11.88


## Conversion Rate Analysis by Page and Treatment Group

This query analyzes conversion performance based on page type and treatment group.

The page column contains:
- old_page
- new_page

The con_treat column contains:
- control group
- treatment group

COUNT(*) calculates the total number of users for each page and treatment combination.

SUM(converted) calculates the total number of converted users.

AVG(converted) calculates the conversion rate because:
- 1 = converted
- 0 = not converted

The conversion rate is multiplied by 100 and rounded to 2 decimal places to display a percentage.

GROUP BY page, con_treat separates the analysis by:
- page type
- treatment group

ORDER BY conversion_rate DESC sorts the results from highest to lowest conversion rate.

This analysis helps determine whether the new page improved user conversion compared to the old page.

Page-level analysis is important in A/B testing because it helps evaluate the impact of design changes, user experience improvements, and product experiments on business performance.

In [48]:
# Perform SQL query to calculate the percentage uplift in conversion rate for the treatment group compared to the control group
pd.read_sql("""
SELECT
    ROUND(
        (
            MAX(CASE WHEN con_treat = 'treatment'
                THEN conversion_rate END)
            -
            MAX(CASE WHEN con_treat = 'control'
                THEN conversion_rate END)
        )
        /
        MAX(CASE WHEN con_treat = 'control'
            THEN conversion_rate END)
        * 100,
    2) AS uplift_percentage
FROM (
    SELECT
        con_treat,
        AVG(converted) * 100 AS conversion_rate
    FROM ab_test
    GROUP BY con_treat
) x;
""", engine)


,uplift_percentage
0,-1.23


## Uplift Analysis

The uplift percentage was -1.23%.

This means the treatment page reduced conversion by 1.23% compared to the control page.

A negative uplift indicates that the treatment version did not improve performance.

P-value / Statistical Significance

In [49]:
from statsmodels.stats.proportion import proportions_ztest

# converted users
converted = [17723, 17514]

# total users
total = [147202, 147276]

# p-value test
z_stat, p_value = proportions_ztest(converted, total)

print("P-value:", round(p_value, 4))

P-value: 0.2161


## Statistical Significance

The p-value was 0.2161.

Since the p-value is greater than 0.05, the result is not statistically significant.

This means there is not enough evidence to say that the treatment page performed differently from the control page.

In [50]:
# Create a SQL view to simplify analysis of A/B test results by country and treatment group
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("DROP VIEW IF EXISTS vw_ab_test_analysis;"))

    conn.execute(text("""
        CREATE VIEW vw_ab_test_analysis AS
        SELECT
            c.country,
            a.page,
            a.con_treat,
            a.converted,
            a.minutes
        FROM ab_test a
        JOIN countries_ab c
        ON a.id = c.id
    """))

In [51]:
# Verify the view was created successfully
pd.read_sql("""
SHOW FULL TABLES
WHERE Table_type = 'VIEW';
""", engine)

,Tables_in_fact_ab_test,Table_type
0,vw_ab_test_analysis,VIEW


In [52]:
# Query the view to verify it works correctly
pd.read_sql("SELECT * FROM vw_ab_test_analysis LIMIT 5;", engine)

,country,page,con_treat,converted,minutes
0,US,old_page,control,0,15.040000
1,US,old_page,control,0,1.453333
2,US,new_page,treatment,0,53.046667
3,US,new_page,treatment,0,57.643333
4,US,new_page,treatment,0,4.746667
